### PR TITLE
libpod: cleanupNetwork() return error

### PR DESCRIPTION
**Pull request by Luap99**
Return the error not just log as the caller can then decide to log this and exit > 0. I also removed the c.valid check as I do not see what the purpose of this would be. c.valid is only false when the ctr was removed but then we should never get there as Cleanup() will not work on a container in removing state.

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by Luap99** [comment]
@mheon PTAL

**Comment by mheon** [comment]
LGTM

**Comment by Luap99** [comment]
@edsantiago please pick this into the no flake try PRs, I have not seen the flake in https://github.com/containers/podman/pull/23487 for 3 pushes now but I am not sure if this is just luck because I don't think  this is actually changing anything other than the error message.

**Comment by edsantiago** [comment]
https://api.cirrus-ci.com/v1/artifact/task/4937874783600640/html/int-podman-rawhide-rootless-host-sqlite.log.html#t--Podman-kube-play-test-env-value-from-configmap-and-replace-should-reuse-the-configmap-volume--1

**Comment by edsantiago** [comment]
Also this [hang in debian](https://api.cirrus-ci.com/v1/artifact/task/6485987155509248/html/int-remote-debian-13-root-host-sqlite.log.html#t--Podman-kube-generate-privileged-container--1) is a new one I don't recall having seen before, don't know if it's related to any recent work

**Comment by Luap99** [comment]
Ack I saw a [hang](https://api.cirrus-ci.com/v1/artifact/task/6388316566716416/html/int-remote-fedora-40-root-host-sqlite.log.html#t--Podman-kube-play-test-with-annotation-size-within-limits--1) on my PR as well but I cannot see how this change would cause this.

**Comment by edsantiago** [comment]
that's #22246, the one Matt is working on

**Comment by Luap99** [comment]
> that's #22246, the one Matt is working on

Right but #22246 hangs on pod rm --all not on stop --all which you linked which is the bit that concerns me here.

File: libpod/container_internal_freebsd.go [hunk 1]
```diff
@@ -146,15 +146,18 @@ func (c *Container) cleanupNetwork() error {
...
-	}
+	neterr := c.runtime.teardownNetNS(c)
 
-	if c.valid {
-		return c.save()
+	// always safe even when there was an error
+	err = c.save()
+	if err != nil {
+		if neterr != nil {
+			logrus.Errorf("Unable to clean up network for container %s: %q", c.ID(), err)
```
Comment (Luap99) [inline libpod/container_internal_freebsd.go:155]: this is logging the wrong error

File: libpod/container_internal_freebsd.go [hunk 2]
```diff
@@ -146,15 +146,18 @@ func (c *Container) cleanupNetwork() error {
...
 
 	// Stop the container's network namespace (if it has one)
-	if err := c.runtime.teardownNetNS(c); err != nil {
-		logrus.Errorf("Unable to cleanup network for container %s: %q", c.ID(), err)
-	}
+	neterr := c.runtime.teardownNetNS(c)
 
-	if c.valid {
-		return c.save()
+	// always safe even when there was an error
```
Comment (rhatdan) [inline libpod/container_internal_freebsd.go:151]: nit s/safe/save?

File: libpod/container_internal_linux.go [hunk 3]
```diff
@@ -187,18 +187,20 @@ func (c *Container) cleanupNetwork() error {
...
-		logrus.Errorf("Unable to clean up network for container %s: %q", c.ID(), err)
-	}
-
+	neterr := c.runtime.teardownNetNS(c)
 	c.state.NetNS = ""
 	c.state.NetworkStatus = nil
 
-	if c.valid {
-		return c.save()
+	// always safe even when there was an error
```
Comment (rhatdan) [inline libpod/container_internal_linux.go:194]: safe or save?

**Comment by edsantiago** [comment]
Is [this](https://api.cirrus-ci.com/v1/artifact/task/5314406916030464/html/int-podman-fedora-39-rootless-host-boltdb.log.html#t--Podman-UserNS-support-podman-userns-container-CTR--1) what you were hoping for?
```
$ podman [options] stop --all -t 0
Error: removing container 953ccd87aeda89841fa7a2e00020d84c78d74a07d67b6983cce00c6e7572ade5 network: unmounting network namespace for container 953ccd87aeda89841fa7a2e00020d84c78d74a07d67b6983cce00c6e7572ade5: failed to remove ns path: remove /run/user/3445/netns/netns-c979509f-2f3d-b0f5-3e2b-3421610b3fe1: device or resource busy
```

**Comment by edsantiago** [comment]
Or [this one](https://api.cirrus-ci.com/v1/artifact/task/6012645789138944/html/int-podman-fedora-40-root-host-sqlite.log.html#t--Podman-pod-start-podman-pod-start-single-pod-by-name--1) but I don't see the new error message here either

**Comment by Luap99** [comment]
> Is [this](https://api.cirrus-ci.com/v1/artifact/task/5314406916030464/html/int-podman-fedora-39-rootless-host-boltdb.log.html#t--Podman-UserNS-support-podman-userns-container-CTR--1) what you were hoping for?
> 
> ```
> $ podman [options] stop --all -t 0
> Error: removing container 953ccd87aeda89841fa7a2e00020d84c78d74a07d67b6983cce00c6e7572ade5 network: unmounting network namespace for container 953ccd87aeda89841fa7a2e00020d84c78d74a07d67b6983cce00c6e7572ade5: failed to remove ns path: remove /run/user/3445/netns/netns-c979509f-2f3d-b0f5-3e2b-3421610b3fe1: device or resource busy
> ```

yes that one should be fixed with the fixes in c/common I think https://github.com/containers/common/pull/2112  https://github.com/containers/podman/pull/23519

But that doesn't fix the very clear issue of podman calling cleanup twice.

**Pull request merged**